File: src/clients/pool.py
"""
PhantomDB Connection Pool for Python

This module provides connection pooling functionality for PhantomDB clients.
"""

import threading
import time
from queue import Queue, Empty
from typing import Optional, Dict, Any
from .phantomdb import PhantomDB, PhantomDBError

class PhantomDBConnectionPool:
    """
    A connection pool for PhantomDB clients.
    
    This class manages a pool of PhantomDB connections to improve performance
    by reusing connections instead of creating new ones for each operation.
    """
    
    def __init__(self, base_url: str = "http://localhost:8080", 
                 min_connections: int = 2, max_connections: int = 10,
                 acquire_timeout: int = 30, idle_timeout: int = 60):
        """
        Initialize the connection pool.
        
        Args:
            base_url: The base URL of the PhantomDB REST API server
            min_connections: Minimum number of connections in the pool
            max_connections: Maximum number of connections in the pool
            acquire_timeout: Timeout for acquiring a connection (seconds)
            idle_timeout: Timeout for idle connections (seconds)
        """
        self.base_url = base_url
        self.min_connections = min_connections
        self.max_connections = max_connections
        self.acquire_timeout = acquire_timeout
        self.idle_timeout = idle_timeout
        
        # Pool state
        self._pool = Queue()
        self._used = set()
        self._lock = threading.Lock()
        self._closed = False
        
        # Initialize the pool
        self._initialize_pool()
    
    def _initialize_pool(self) -> None:
        """Initialize the connection pool with minimum connections."""
        for _ in range(self.min_connections):
            self._create_connection()
    
    def _create_connection(self) -> Optional[PhantomDB]:
        """
        Create a new connection and add it to the pool.
        
        Returns:
            PhantomDB client instance or None if at max capacity
        """
        with self._lock:
            if self._pool.qsize() + len(self._used) >= self.max_connections:
                return None
            
            connection = PhantomDB(self.base_url)
            self._pool.put({
                'client': connection,
                'last_used': time.time(),
                'idle': True
            })
            
            return connection
    
    def acquire(self) -> PhantomDB:
        """
        Acquire a connection from the pool.
        
        Returns:
            PhantomDB client instance
            
        Raises:
            PhantomDBError: If unable to acquire a connection within the timeout
        """
        if self._closed:
            raise PhantomDBError("Connection pool is closed")
        
        start_time = time.time()
        
        while time.time() - start_time < self.acquire_timeout:
            try:
                # Try to get an available connection
                connection_info = self._pool.get_nowait()
                
                # Check if connection is still valid
                if time.time() - connection_info['last_used'] < self.idle_timeout:
                    connection_info['idle'] = False
                    with self._lock:
                        self._used.add(id(connection_info['client']))
                    return connection_info['client']
                # Connection is too old, discard it and try again
            except Empty:
                # No available connections, check if we can create a new one
                with self._lock:
                    if len(self._used) < self.max_connections:
                        connection = self._create_connection()
                        if connection:
                            connection_info = {
                                'client': connection,
                                'last_used': time.time(),
                                'idle': False
                            }
                            with self._lock:
                                self._used.add(id(connection_info['client']))
                            return connection
                
                # Wait a bit before trying again
                time.sleep(0.1)
        
        raise PhantomDBError("Timeout acquiring connection from pool")
    
    def release(self, connection: PhantomDB) -> None:
        """
        Release a connection back to the pool.
        
        Args:
            connection: The PhantomDB client instance to release
        """
        if self._closed:
            return
        
        with self._lock:
            connection_id = id(connection)
            if connection_id in self._used:
                self._used.remove(connection_id)
                
                # Add back to pool
                self._pool.put({
                    'client': connection,
                    'last_used': time.time(),
                    'idle': True
                })
    
    def close(self) -> None:
        """Close the connection pool and all connections."""
        self._closed = True
        
        # Clear the pool
        while not self._pool.empty():
            try:
                self._pool.get_nowait()
            except Empty:
                break
        
        with self._lock:
            self._used.clear()
    
    def get_status(self) -> Dict[str, Any]:
        """
        Get the current pool status.
        
        Returns:
            Dictionary with pool status information
        """
        with self._lock:
            return {
                'available': self._pool.qsize(),
                'used': len(self._used),
                'min': self.min_connections,
                'max': self.max_connections
            }

def create_pool(base_url: str = "http://localhost:8080", 
                min_connections: int = 2, max_connections: int = 10,
                acquire_timeout: int = 30, idle_timeout: int = 60) -> PhantomDBConnectionPool:
    """
    Create a new PhantomDB connection pool.
    
    Args:
        base_url: The base URL of the PhantomDB REST API server
        min_connections: Minimum number of connections in the pool
        max_connections: Maximum number of connections in the pool
        acquire_timeout: Timeout for acquiring a connection (seconds)
        idle_timeout: Timeout for idle connections (seconds)
        
    Returns:
        A new connection pool instance
    """
    return PhantomDBConnectionPool(base_url, min_connections, max_connections, 
                                 acquire_timeout, idle_timeout)

File: src/clients/javascript/pool.js
/**
 * PhantomDB Connection Pool for JavaScript
 * 
 * This module provides connection pooling functionality for PhantomDB clients.
 */

class PhantomDBConnectionPool {
  /**
   * Create a new connection pool.
   * 
   * @param {Object} options - Pool configuration options
   * @param {string} options.baseUrl - The base URL of the PhantomDB REST API server
   * @param {number} [options.minConnections=2] - Minimum number of connections in the pool
   * @param {number} [options.maxConnections=10] - Maximum number of connections in the pool
   * @param {number} [options.acquireTimeout=30000] - Timeout for acquiring a connection (ms)
   * @param {number} [options.idleTimeout=60000] - Timeout for idle connections (ms)
   */
  constructor(options = {}) {
    this.baseUrl = options.baseUrl || 'http://localhost:8080';
    this.minConnections = options.minConnections || 2;
    this.maxConnections = options.maxConnections || 10;
    this.acquireTimeout = options.acquireTimeout || 30000;
    this.idleTimeout = options.idleTimeout || 60000;
    
    // Pool state
    this.pool = [];
    this.used = new Set();
    this.pending = [];
    this.closed = false;
    
    // Initialize the pool
    this._initializePool();
  }
  
  /**
   * Initialize the connection pool with minimum connections.
   * @private
   */
  _initializePool() {
    for (let i = 0; i < this.minConnections; i++) {
      this._createConnection();
    }
  }
  
  /**
   * Create a new connection and add it to the pool.
   * @private
   */
  _createConnection() {
    if (this.pool.length + this.used.size >= this.maxConnections) {
      return null;
    }
    
    const { PhantomDB } = require('./phantomdb');
    const connection = new PhantomDB(this.baseUrl);
    this.pool.push({
      client: connection,
      lastUsed: Date.now(),
      idle: true
    });
    
    return connection;
  }
  
  /**
   * Acquire a connection from the pool.
   * 
   * @returns {Promise<Object>} A PhantomDB client instance
   * @throws {Error} If unable to acquire a connection within the timeout
   */
  async acquire() {
    if (this.closed) {
      throw new Error('Connection pool is closed');
    }
    
    // Try to get an available connection
    while (this.pool.length > 0) {
      const connectionInfo = this.pool.pop();
      if (Date.now() - connectionInfo.lastUsed < this.idleTimeout) {
        connectionInfo.idle = false;
        this.used.add(connectionInfo);
        return connectionInfo.client;
      }
      // Connection is too old, discard it
    }
    
    // If we're at max capacity, wait for a connection to be released
    if (this.used.size >= this.maxConnections) {
      return new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
          reject(new Error('Timeout acquiring connection from pool'));
        }, this.acquireTimeout);
        
        this.pending.push({ resolve, reject, timeout });
      });
    }
    
    // Create a new connection
    const connection = this._createConnection();
    if (connection) {
      const connectionInfo = this.pool.find(info => info.client === connection);
      if (connectionInfo) {
        connectionInfo.idle = false;
        this.used.add(connectionInfo);
      }
      return connection;
    }
    
    throw new Error('Unable to acquire connection');
  }
  
  /**
   * Release a connection back to the pool.
   * 
   * @param {Object} connection - The PhantomDB client instance to release
   */
  release(connection) {
    if (this.closed) {
      return;
    }
    
    // Find the connection info
    const connectionInfo = Array.from(this.used).find(info => info.client === connection);
    if (connectionInfo) {
      // Remove from used set
      this.used.delete(connectionInfo);
      
      // Update last used time
      connectionInfo.lastUsed = Date.now();
      connectionInfo.idle = true;
      
      // Add back to pool
      this.pool.push(connectionInfo);
      
      // Check if there are pending requests
      if (this.pending.length > 0) {
        const pending = this.pending.shift();
        clearTimeout(pending.timeout);
        this.acquire().then(pending.resolve).catch(pending.reject);
      }
    }
  }
  
  /**
   * Close the connection pool and all connections.
   */
  async close() {
    this.closed = true;
    
    // Clear pending requests
    while (this.pending.length > 0) {
      const pending = this.pending.shift();
      clearTimeout(pending.timeout);
      pending.reject(new Error('Connection pool closed'));
    }
    
    // Clear the pool
    this.pool = [];
    this.used.clear();
  }
  
  /**
   * Get the current pool status.
   * 
   * @returns {Object} Pool status information
   */
  getStatus() {
    return {
      available: this.pool.length,
      used: this.used.size,
      pending: this.pending.length,
      min: this.minConnections,
      max: this.maxConnections
    };
  }
}

/**
 * Create a new PhantomDB connection pool.
 * 
 * @param {Object} options - Pool configuration options
 * @returns {PhantomDBConnectionPool} A new connection pool instance
 */
function createPool(options) {
  return new PhantomDBConnectionPool(options);
}

module.exports = {
  PhantomDBConnectionPool,
  createPool
};